请问模型裁剪只能针对yolov3么？

如果想对faster-rcnn的模型进行裁剪，应该怎么做呢？

谢谢！

> 
> 
> 首先，参考这个教程计算敏感度：
> https://github.com/PaddlePaddle/PaddleDetection/tree/release/0.2/slim/sensitive
> 
> 然后，参考这个教程，将其中的配置文件选项设置为faster-rcnn配置文件即可。
> https://github.com/PaddlePaddle/PaddleDetection/tree/release/0.2/slim/prune

按照这个操作计算敏感度，然后prune都没问题，但是prune之后eval的时候报错了。AssertionError: Shape not matching: the Program requires a parameter with a shape of ((128,)), while the loaded parameter (namely [ bn3a_branch2b_mean ]) has a shape of  ((30,)).

> 
> 
> ```
> python eval.py \
> -c ../../configs/yolov3_mobilenet_v1_voc.yml \
> --pruned_params "yolo_block.0.0.0.conv.weights,yolo_block.0.0.1.conv.weights,yolo_block.0.1.0.conv.weights" \
> --pruned_ratios="0.2,0.3,0.4" \
> -o weights=output/yolov3_mobilenet_v1_voc/model_final
> ```
> 
> 请确认下`--pruned_params`和`--pruned_ratios`设置的和训练时一致？

一致的。刚用了示例的参数重跑了，还是一样的报错，日志如下：

python prune.py -c G:\workspace\github\paddle\PaddleDetection\configs\test\yolov3_mobilenet_v1_p.yml --pruned_params "yolo_block.0.0.0.conv.weights,yolo_block.0.0.1.conv.weights,yolo_block.0.1.0.conv.weights" --pruned_ratios="0.2,0.3,0.4"

2020-03-18 11:35:46,569-INFO: Save model to I:/test/models/output_prune\yolov3_mobilenet_v1_p\model_final.

python eval.py -c G:\workspace\github\paddle\PaddleDetection\configs\test\yolov3_mobilenet_v1_p.yml --pruned_params "yolo_block.0.0.0.conv.weights,yolo_block.0.0.1.conv.weights,yolo_block.0.1.0.conv.weights" --pruned_ratios="0.2,0.3,0.4" -o weights=I:/test/models/output_prune/yolov3_mobilenet_v1_p/model_final

Traceback (most recent call last):
  File "eval.py", line 229, in <module>
    main()
  File "eval.py", line 179, in main
    checkpoint.load_params(exe, eval_prog, cfg.weights)
  File "G:\workspace\github\paddle\PaddleDetection\ppdet\utils\checkpoint.py", line 141, in load_params
    fluid.io.set_program_state(prog, state)
  File "C:\Users\admin\AppData\Roaming\Python\Python36\site-packages\paddle\fluid\io.py", line 1942, in set_program_state
    .format(orig_para_np.shape, para.name, new_para_np.shape)
AssertionError: Shape not matching: the Program requires a parameter with a shape of ((512,)), while the loaded parameter (namely [ yolo_block.0.1.0.bn.mean ]) has a shape of  ((307,)).

@wanghaoshuang 有。完整的是这样的。
2020-03-20 16:27:59,260-INFO: pruned params: ['yolo_block.0.0.0.conv.weights', 'yolo_block.0.0.1.conv.weights', 'yolo_block.0.1.0.conv.weights']
2020-03-20 16:27:59,261-INFO: pruned ratios: [0.2, 0.3, 0.4]
2020-03-20 16:27:59,437-INFO: pruned FLOPS: 0.18126663050258127
W0320 16:28:02.881803 20184 device_context.cc:237] Please NOTE: device: 0, CUDA Capability: 61, Driver API Version: 10.0, Runtime API Version: 9.0
W0320 16:28:07.694550 20184 device_context.cc:245] device: 0, cuDNN Version: 7.4.
2020-03-20 16:28:09,742-INFO: Loading parameters from I:/te_020/models/output_prune/yolov3_mobilenet_v1_p/model_final...
Traceback (most recent call last):
File "eval.py", line 229, in <module>
File "eval.py", line 179, in main
checkpoint.load_params(exe, eval_prog, cfg.weights)
File "G:\workspace\github\paddle\PaddleDetection\ppdet\utils\checkpoint.py", line 141, in load_params
fluid.io.set_program_state(prog, state)
File "C:\Users\admin\AppData\Roaming\Python\Python36\site-packages\paddle\fluid\io.py", line 1942, in set_program_state
.format(orig_para_np.shape, para.name, new_para_np.shape)
AssertionError: Shape not matching: the Program requires a parameter with a shape of ((512,)), while the loaded parameter (namely [ yolo_block.0.1.0.bn.mean ]) has a shape of ((307,)).

@wanghaoshuang 有。完整的是这样的。

`2020-03-20 16:27:59,260-INFO: pruned params: ['yolo_block.0.0.0.conv.weights', 'yolo_block.0.0.1.conv.weights', 'yolo_block.0.1.0.conv.weights']
2020-03-20 16:27:59,261-INFO: pruned ratios: [0.2, 0.3, 0.4]
2020-03-20 16:27:59,437-INFO: pruned FLOPS: 0.18126663050258127
W0320 16:28:02.881803 20184 device_context.cc:237] Please NOTE: device: 0, CUDA Capability: 61, Driver API Version: 10.0, Runtime API Version: 9.0
W0320 16:28:07.694550 20184 device_context.cc:245] device: 0, cuDNN Version: 7.4.
2020-03-20 16:28:09,742-INFO: Loading parameters from I:/test/models/output_prune/yolov3_mobilenet_v1_p/model_final...
Traceback (most recent call last):
File "eval.py", line 229, in 
File "eval.py", line 179, in main
checkpoint.load_params(exe, eval_prog, cfg.weights)
File "G:\workspace\github\paddle\PaddleDetection\ppdet\utils\checkpoint.py", line 141, in load_params
fluid.io.set_program_state(prog, state)
File "C:\Users\admin\AppData\Roaming\Python\Python36\site-packages\paddle\fluid\io.py", line 1942, in set_program_state
.format(orig_para_np.shape, para.name, new_para_np.shape)
AssertionError: Shape not matching: the Program requires a parameter with a shape of ((512,)), while the loaded parameter (namely [ yolo_block.0.1.0.bn.mean ]) has a shape of ((307,)).`

--- a/slim/prune/eval.py
+++ b/slim/prune/eval.py
@@ -86,6 +86,7 @@ def main():
                 fetches = model.eval(feed_vars, multi_scale_test)
     eval_prog = eval_prog.clone(True)
 
+    exe.run(startup_prog)
     reader = create_reader(cfg.EvalReader)
     loader.set_sample_list_generator(reader, place)
 
@@ -123,7 +124,7 @@ def main():
         params=pruned_params,
         ratios=pruned_ratios,
         place=place,
-        only_graph=True)
+        only_graph=False)
     pruned_flops = flops(eval_prog)
     logger.info("pruned FLOPS: {}".format(
         float(base_flops - pruned_flops) / base_flops))
@@ -174,7 +175,6 @@ def main():
         sub_eval_prog = sub_eval_prog.clone(True)
 
     # load model
-    exe.run(startup_prog)
     if 'weights' in cfg:
         checkpoint.load_checkpoint(exe, eval_prog, cfg.weights)
 

